d3a_interface/matching_algorithms/requirements_validators.py

```diff
@@ -0,0 +1,205 @@
+"""
+Copyright 2018 Grid Singularity
+This file is part of D3A.
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+
+from abc import ABC, abstractmethod
+from typing import Dict, Optional, Union
+
+from d3a_interface.data_classes import Offer, Bid
+
+
+class Requirement(ABC):
+    """Interface for offer and bid requirements."""
+
+    @classmethod
+    @abstractmethod
+    def is_satisfied(
+            cls, offer: Offer, bid: Bid, requirement: Dict,
+            clearing_rate: Optional[float] = None,
+            selected_energy: Optional[float] = None) -> bool:
+        """Check whether a requirement is satisfied."""
+
+
+class TradingPartnersRequirement(Requirement):
+    """Check if trading_partners requirement is satisfied for both bid and offer."""
+
+    @classmethod
+    def is_satisfied(cls, offer: Offer, bid: Bid, requirement: Dict,
+                     clearing_rate: Optional[float] = None,
+                     selected_energy: Optional[float] = None) -> bool:
```

Comment:
I'm a bit confused by this structure. This class is a `Requirement`, but this method has an argument named `requirement`. I get that they're different objects, but the terminology is a bit confusing and it's hiding this difference, making it difficult to understand what's going on.

Besides that: pylint complains that this method has "too many arguments". It might be a hint for a possible refactoring. Also: please make sure to add docstrings to all public methods (we discussed this in Lisbon and then on slack) :)

Reply:
I also recall that we agreed not to make a big deal of doctsrings that only show `what is done`
Anyway, I did add a docstring for it in the base class interface method `is_satisfied` so I don't think that I should add docstring each time I implement this method :)

Furthermore, it's true that the class is `Requirement` but the method is abstract, so we do need to pass a requirement to check whether it's satisfied or not

Reply:
In fact, `clearing_rate` and `selected_energy` are not used for this method, nor for any of the other `Requirement` classes. The only one that uses it is `PriceRequirement`, but I think that adding these parameters to all other classes just for this reason isn't the best option. I have the feeling that we could still improve this interface.

Reply:
A requirement needs extra context which consists of both the offer / bid, the selected energy between them and the decided on clearing rate
I made the interface accept them all because I suspect these are the only stakeholders at the moment
In general I'm okay with further refactoring the interface

Reply:
> the selected energy between them and the decided on clearing rate

But as far as I see, the classes that implement `Requirement` don't use all of the arguments. That's why I think the interface should be changed.

Reply:
@fievelk you are right, It's worth giving the design another look

Reply:
It may not be the `perfect` design yet, however the main rational behind it was to make sure it's scalable to other requirements that could be added in the future

Reply:
> It may not be the perfect design yet, however the main rational behind it was to make sure it's scalable to other requirements that could be added in the future

The problem is exactly that every time we add a new requirement (with its sets of arguments) we need to modify all existing methods to also accept those parameters. That's why I think this is not very scalable. Anyway, this can be dealt with it in the other ticket.